From e9068c5b3c0211cfeb8a276b7590b7afa8c59c59 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:54:41 -0400
Subject: [fs] jbd2: check jh->b_transaction without taking j_list_lock

Message-id: <1414058125-4183-128-git-send-email-lczerner@redhat.com>
Patchwork-id: 98261
O-Subject: [RHEL 7.1 PATCH v2 127/171] jbd2: check jh->b_transaction without taking j_list_lock
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Theodore Ts'o <tytso@mit.edu>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit d2eb0b998990abf51d6e1d3bf16a2637b920a660

 jh->b_transaction is adequately protected for reading by the
 jbd_lock_bh_state(bh), so we don't need to take j_list_lock in
 __journal_try_to_free_buffer().

 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/jbd2/transaction.c b/fs/jbd2/transaction.c
index 8360674..902b99b 100644
--- a/fs/jbd2/transaction.c
+++ b/fs/jbd2/transaction.c
@@ -1819,11 +1819,11 @@ __journal_try_to_free_buffer(journal_t *journal, struct buffer_head *bh)
  if (buffer_locked(bh) || buffer_dirty(bh))
   goto out;
 
- if (jh->b_next_transaction != NULL)
+ if (jh->b_next_transaction != NULL || jh->b_transaction != NULL)
   goto out;
 
  spin_lock(&journal->j_list_lock);
- if (jh->b_cp_transaction != NULL && jh->b_transaction == NULL) {
+ if (jh->b_cp_transaction != NULL) {
   /* written-back checkpointed metadata buffer */
   JBUFFER_TRACE(jh, "remove from checkpoint list");
   __jbd2_journal_remove_checkpoint(jh);
-- 
1.7.1